temporalio/envconfig.py

```diff
@@ -0,0 +1,342 @@
+            return f.read()
+    return source
+
+
+@dataclass(frozen=True)
+class ClientConfigTls:
+    """TLS configuration as specified as part of client configuration
+
+    .. warning::
+        Experimental API.
+    """
+
+    disabled: bool = False
+    """If true, TLS is explicitly disabled."""
+    server_name: Optional[str] = None
+    """SNI override."""
+    server_root_ca_cert: Optional[DataSource] = None
+    """Server CA certificate source."""
+    client_cert: Optional[DataSource] = None
+    """Client certificate source."""
+    client_private_key: Optional[DataSource] = None
+    """Client key source."""
+
+    def to_connect_tls_config(self) -> Union[bool, TLSConfig]:
+        """Create a `temporalio.service.TLSConfig` from this profile."""
+        if self.disabled:
+            return False
+
+        return TLSConfig(
+            domain=self.server_name,
+            server_root_ca_cert=_read_source(self.server_root_ca_cert),
+            client_cert=_read_source(self.client_cert),
+            client_private_key=_read_source(self.client_private_key),
+        )
+
+    @staticmethod
+    def _from_dict(d: Optional[Mapping[str, Any]]) -> Optional[ClientConfigTls]:
+        if not d:
+            return None
+        return ClientConfigTls(
+            disabled=d.get("disabled", False),
+            server_name=d.get("server_name"),
+            # Note: Bridge uses snake_case, but TOML uses kebab-case which is
+            # converted to snake_case. Core has server_ca_cert, client_key.
+            server_root_ca_cert=_from_dict_to_source(d.get("server_ca_cert")),
+            client_cert=_from_dict_to_source(d.get("client_cert")),
+            client_private_key=_from_dict_to_source(d.get("client_key")),
+        )
+
+
+@dataclass(frozen=True)
+class ClientConfigProfile:
+    """Represents a client configuration profile.
+
+    This class holds the configuration as loaded from a file or environment.
+    See `to_connect_config` to transform the profile to `temporalio.service.ConnectConfig`,
+    which can be used to create a client.
+
+    .. warning::
+        Experimental API.
+    """
+
+    address: Optional[str] = None
+    """Client address."""
+    namespace: Optional[str] = None
+    """Client namespace."""
+    api_key: Optional[str] = None
+    """Client API key."""
+    tls: Optional[ClientConfigTls] = None
+    """TLS configuration."""
+    grpc_meta: Mapping[str, str] = field(default_factory=dict)
+    """gRPC metadata."""
+
+    @staticmethod
+    def from_dict(d: Mapping[str, Any]) -> ClientConfigProfile:
+        """Create a ClientConfigProfile from a dictionary from the bridge."""
+        return ClientConfigProfile(
+            address=d.get("address"),
+            namespace=d.get("namespace"),
+            api_key=d.get("api_key"),
+            tls=ClientConfigTls._from_dict(d.get("tls")),
+            grpc_meta=d.get("grpc_meta") or {},
+        )
+
+    def to_connect_config(self) -> ConnectConfig:
```

Comment:
Can you give an example of the ergonomics of how this looks from a simple user/sample POV? At https://github.com/temporalio/proposals/blob/master/all-sdk/external-client-configuration.md#python-sdk-idea we had:

```python
config = envconfig.ClientConfig.load_client_connect_config()
client = await Client.connect(**config)
```

So sounds like we need a helper at `load_client_connect_config` that is the equivalent of https://pkg.go.dev/go.temporal.io/sdk/contrib/envconfig#LoadClientOptions and/or https://pkg.go.dev/go.temporal.io/sdk/contrib/envconfig#LoadDefaultClientOptions?

Reply:
:+1: Will comment on that signature specifically